### ISSUE TITLE
ExternalIP can be abused in some network configurations

### ISSUE THREAD

**Issue by smarterclayton** [kubernetes/kubernetes#22650]
Depending on the cluster network configuration, it's possible for an admin to not realize that service `externalIP` can be used to trick nodes into diverting incoming traffic to a service under a different user's control.  As an example, an SDN implementation that delivers packets over the SDN onto a node doesn't have the same iptables conditions as our existing externalIP checks, which means that packets routed from pods to nodes (for a node port or service fronting pods using hostNetwork) will accept the external IP rule.

We should give the administrator control over which externalIPs are allowed to be set or whether externalIP can be set at all.  Using one or more CIDR ranges is sufficient (admin can choose `0.0.0.0/0`).   In the future, we probably want to have externalIPs be not directly settable by end cluster users, since they have the potential to allow any cluster user to masquerade as any other service that is also using externalIP

**Comment by smarterclayton** [comment]
@kubernetes/sig-network

**Comment by eparis** [comment]
This does not make sense to me. What exactly are you saying is wrong here? @danwinship was unable to find any case of externalIP that didn't do exactly what it was supposed to do...

**Comment by danwinship** [comment]
You can use the address of a node as an ExternalIP, and thus steal traffic from a particular (unused) port on a node. While that is more or less "exactly what it was supposed to do", it's not a very good behavior and probably *shouldn't* work.

(Note that you can also do the same thing with `hostPort` on a pod, which is also both exactly what it was supposed to do, and not a good idea in general.)

**Comment by eparis** [comment]
@danwinship sounds to me like you are saying 'ExternalIP' works exactly as it is supposed to work. (Not asking if its a good idea, but it is doing exactly what it says it will do)

There are definitely things we might want to do to make this better....  Do we want to enforce the NodePort port restrictions on ExternalIPs? So an ExternalIP can't race in and grab port 22 while sshd is restarting?

@thockin @mikedanese

**Comment by eparis** [comment]
I don't believe `a` CIDR makes sense. These are not virtual resources that we can enforce contiguous-ness like other virtual network concepts (cluster ip, service ip, etc). These are addresses that live in the 'real' network. So at a minimum this would have to be a list of CIDRs. But once we are talking about a list of CIDR it starts to look more like API objects rather than command line arguments...

And even then what are we protecting?

**Comment by smarterclayton** [comment]
> You can use the address of a node as an ExternalIP, and thus steal traffic from a particular (unused) port on a node.

An unused port may very well be a port that was used until just recently, but a rolling update of a daemon set or RC is in progress, and the pod disappears and releases its port.  For a period of time before a new pod is scheduled, the other service can harvest connections from that unused port.  Those connections might be port 80 for a load balancer you're running on hostNetwork, etc.

> And even then what are we protecting?

Agree a single CIDR doesn't make sense, unless we're using externalIP as a virtual block that we're dividing up between services, and that case won't work in practice because anyone can grab any IP on in the block, even if someone else already grabbed it.  We'd probably need a redesign of the externalIP function to make them practically secure in the presence of malicious clients, but in the scenario where an admin is already leveraging them a single on/off wouldn't work.

**Comment by thockin** [comment]
Circling back to old issues in my inbox.  Let it be knows that I used to
not like this feature.  I still don't, but I used to, also.  In reality it
is very niche and rather weird in its behavior.  But I was convinced to
keep it because it was useful for some important cases.  Having a way to
disable it would be OK to me.

**Comment by kiall** [comment]
This does seem like it can be abused to hijack traffic from other workloads - i.e. a possible security issue. @eparis's example of hijacking the hosts SSH port seems plausible, as do @smarterclayton's examples of hijacking other workloads ports.

**Comment by kiall** [comment]
I spent some time trying this out. Hijacking other workloads was trivial. Create a external IP service for the right port, restart kube-proxy once, and I had the traffic.

I don't know if this was a lucky coincidence or repeatable. I.e. I don't know if kube-proxy is binding external IP services before NodePort services, or if I just "won" with a race condition.

For host services, like SSH, I managed to grab that port too - but it was significantly harder.. I needed to restart SSHD either slowly (stop, wait 5 mins, start) or repeatably restart, so that kube-proxy would attempt the bind in the few milliseconds SSHD wasn't holding the port itself.

And, given the external-IP bind was more specific (it bound to the nodes actual IP), and SSHD was configured for 0.0.0.0:22 - once kube-proxy had the more specific bind, SSHD would start up normally, and successfully get the less specific 0.0.0.0:22 bind.
